### PR TITLE
chunk refactor experiment

### PR DESCRIPTION
**Pull request by briandorsey**
Attempting to combine advice from [kitlith](https://github.com/kitlith) in #25 and comments on 8613798e76dde32d1d2c91df414ff60b14086058 into a macro & trait based solution.

### CODE REVIEW

**Comment by briandorsey** [comment]
TODO: 

- [ ] add test for using KnownChunk<something> and then returning it through ChunkT trait. (I think it won't work currently because I'm not implementing ChunkT on KnownChunks properly.) 
- [ ] rename Chunk --> ChunkEnum, then ChunkT --> Chunk ?  (if so, don't forget InfoChunk --> InfoChunkEnum as well)

File: src/lib.rs [hunk 1]
```diff
@@ -1016,18 +1017,28 @@ impl ListInfoChunk {
...
 macro_rules! info_chunks {
     ($(($name:ident,$literal:literal)),*$(,)?) => {
+        paste!{
+        $(
+            #[binrw]
+            #[br(little)]
+            #[derive(Debug, PartialEq, Eq)]
+            pub struct [<$name ChunkData>] {
+                value: NullString,
+            }
```
Comment (kitlith) [inline src/lib.rs:1027]: If you wanted, it feels like you could do the same thing here as I suggested with KnownChunk. i.e. you'd either have `InfoChunk<IENG>` (going the trait based approach but with another new trait) or `InfoChunk<fourcc(b"IENG")>` (going the const generics approach)

Comment (briandorsey) [reply src/lib.rs:1027]: Apologies, I'm probably missing a key point again. For context, this macro is doing two things:

- Generating a series of SOMETHINGData structs & the wrapper KnownChunk<SOMETHINGData> structs. 
- then, creating an enum wrapper around all of those structs + the unknown case. (I renamed the enums to make this clearer)

I didn't succeed in writing the trait version, need to spend some more time with that because I like having the ID as a FourCC instead of u32 for errors/debugging. I did get a const generic approach working: 

```rust
#[binrw]
#[br(little)]
#[derive(Debug, PartialEq, Eq)]
pub struct InfoChunkData<const I: u32> {
    value: NullString,
}

impl<const I: u32> KnownChunkID for InfoChunkData<I> {
    const ID: FourCC = FourCC(I.to_le_bytes());
}

macro_rules! info_chunks {
    ($(($name:ident,$literal:literal)),*$(,)?) => {
        paste!{
        $(
            type [<$name ChunkData>] = InfoChunkData<{ fourcc($literal) }>;
            type [<$name Chunk>] = KnownChunk<[<$name ChunkData>]>;
// ...
```

But it doesn't simplify the macro very much, leading me to thinking that I'm missing your point. ;) 
In summary, I do like the idea of simplifying or even removing this macro, but it doesn't feel like I'm actually on that path at the moment. 

Thinking out loud: Hmm... maybe I just need to stop generating the enum with a macro, ~20 cases (times three matches in trait functions) isn't that many extra lines of code and that might open things up to getting rid of the macro completely.

Comment (briandorsey) [reply src/lib.rs]: OK, I gave that a try, and removed the macro entirely. Const generic for each SOMETHINGData struct and manual creation of the wrapping enum. Overall, I think I like it better. But the ergonomics of the u32 const generics aren't great. The helper function you proposed in a previous comment helps, though! 

Creating one looks like this: 
```rust
InfoChunkData::<{ fourcc(b"ICMT") }> {
                value: NullString("comment".into()),
            }
``` 

Which isn't too bad, I guess, but the macro created version would have looked more common when using it: `IcmtChunkData{ value: ...}`. 

... and... as I was typing that, I realized that I could just add an additional type alias, so now it works like this: 

```rust
 IcmtChunkData {
                value: NullString("comment".into()),
            },
```

Comment (briandorsey) [reply src/lib.rs]: Thanks for the poke on trying generics here! Pretty happy with how this turned out overall.

Comment (kitlith) [reply src/lib.rs]: Whoops, sorry, I wasn't paying attention to github for a few days.

Yeah, I missed the automatic enum creation, there. The generation of type aliases and that enum could probably still be automated with a similar macro, if you wanted.

re: trait based approach: I could take a look at that if you feel like that would give still-better ergonomics over the current solution? Main difference I'd see from how KnownChunk handles it is the requirement of `PhantomData` so that rust doesn't complain about the generic not being used in the type.

Nevertheless, I'm glad you're happy with the current solution!

Comment (briandorsey) [reply src/lib.rs]: While verbose, I think the ergonomics turned out OK in the end. The `KnownChunkID` triait makes things consistent enough for my tastes even though the generic chunks get there a different way.
